[bitnami/redis] Allow the use of NodePort to allow for external access

Maelstrom96 opened the pull request:
<!--
Before you open the request please review the following guidelines and tips to help it be more easily integrated:
 
- Describe the scope of your change - i.e. what the change does.
- Describe any known limitations with your change.
- Please run any tests or examples that can exercise your modified code.
 
Thank you for contributing! We will try to test and integrate the change as soon as we can, but be aware we have many GitHub repositories to manage and can't immediately respond to every request. There is no need to bump or check in on a pull request (it will clutter the discussion of the request).
 
Also don't be worried if the request is closed or not integrated sometimes the priorities of Bitnami might not match the priorities of the pull request. Don't fret, the open source community thrives on forks and GitHub makes it easy to keep your changes in a forked repo.
-->
 
**Description of the change**
 
<!-- Describe the scope of your change - i.e. what the change does. -->
This PR enables Nodeport to allow for a setup where internal and external communication is necessary. This also makes the changes so the instances use hostnames instead of clusterIPs.

Added logic to prevent the Redis service from starting if it’s starting as a slave and sentinels sees its master as down.
 
**Benefits**
 
<!-- What benefits will be realized by the code change? →
Before, when Nodeport was enabled, it didn’t properly configure the services for them to be accessible outside the k8s cluster.
 
With this PR, when Nodeport is enabled, Redis/Sentinel port == Individual Node Service Nodeport.
This allows for internal and external communication to the same endpoints.
 
You need this in order for the Redis clients to get the right connectivity information from the sentinels.
By default, the information sent by sentinel wouldn’t allow for the external client to work properly, since it can’t connect to a ClusterIP. Using service hostnames allows us to define those DNS entries and have them point to any of your Node IP.
 
The Nodeports are programmatically set by the helm chart, or can be manually set.
The automatic setting of ports is the highlight of the PR, but manually setting is supported with some caveats.
 
For automatic port assignment, nodeport is enabled but ports/nodeports are not specified.
A configmap gets created `ports-configmap.yaml` which gets the used nodeports on the k8s cluster via API call https://helm.sh/docs/chart_template_guide/functions_and_pipelines/#using-the-lookup-function
, and generates a list of free nodeports. These ports then get set to the main service as well as new node services. The other configmaps and sentinel statefulset then gets ports by use of envvars.
https://kubernetes.io/docs/concepts/services-networking/service/#environment-variables
 
The automatic port assignment requires an install followed by an upgrade (no Values.yaml change required). Similar to redis-cluster with loadbalancer. This is because the ports-configmap must be installed first, and allowed to gather the ports. Then on the upgrade the services and statefulset are installed. These install/upgrade steps are not needed/used when installing with clusterip, or when ports are manually set.
 
Manually setting of ports is supported. `sentinel.service.nodePorts.redis` and `sentinel.service.nodePorts.sentinel` are used to create the services. The Sentinel and Redis services get the defined ports value, and the nodes services get created by incrementing both ports by one. Therefore, setting these two ports with some distance is required, and it is possible for these manually set ports to collide with already existing Nodeports, failing the chart installation. Manually setting these 2 values may be desired in a firewalled network environment with restricted port access. 
 
An additional benefit is that failovers are improved. A master AND all but one replica can die at the same time, and the cluster will recover automatically (with one caveat). The remaining replica will be promoted by the recovering sentinels ensuring data integrity. A new function was added which prevents new redis replicas from starting until the sentinels form a quorum. This prevents a newly started replica from becoming master, resulting in data on the longest lived replica from being lost. 
 
Fixed sentinel authentication for redis-cli in scripts-configmap.
Changed auth.redis to auth.sentinel since these redis-cli commands are connected to sentinel ports.
 
 
 
**Possible drawbacks**
 
<!-- Describe any known limitations with your change -->
Manually setting nodeports/ports is not as seamless and easy as allowing the automatic generation of them. Right now, `ports-configmap.yaml` is only used for automatic generation of ports.
 
When the automatic ports are generated, it is possible that this collides with one of the null internal nodeports. When this happens, an uninstall/reinstall will fix this.
 
Sometimes because of timings, killing of master+replica may result in a broken cluster. The sentinels fail to elect the remaining Redis replica because Sentinel doesn’t consider the last Replica to be a “GOOD” replica.  Increasing the downAfterMilliseconds alleviates this. . We’ve implemented a failsafe that will try and do a manual failover in the case that the master and sentinel doesn’t failover because the replica is not considered “GOOD”. Without the failsafe, you would have to recover the cluster using the sentinel command `SENTINEL FAILOVER mymaster` on the remaining sentinel node to trigger a manual failover. The rest of the pods should then come back up fine and all the data should still be there.
 
**Applicable issues**
 
<!-- Enter any applicable Issues here (You can reference an issue using #) -->
 - fixes #4082
 - fixes #7283 - Added the failsafe mechanism in case quorum is broken. This issue would also be the result of the value downAfterMilliseconds being too low.
 
**Additional information**
 
<!-- If there's anything else that's important and relevant to your pull
request, mention that information here.-->
If using Calico policies, and possibly policies from other CNIs, kube-proxy needs to use IPVS instead of IPTable.
https://docs.projectcalico.org/networking/enabling-ipvs
 
During the testing of redis sentinel, we ran into an issue where connections from outside the cluster does not behave as desired. Not failing over correctly when the master pod is killed.
Connectivity between the pods always worked, and internal connections from test pods always worked.
 
We discovered it was an issue with iptables not gracefully releasing the connections because of Calico policies. Using IPVS fixed this issue.
 
Additionally, for the external access to work, either hosts file or DNS servers needs to be updated to include hostnames and ips for external routing to work.
 
We have been testing with local helm installs and through Rancher. We have our own repo and catalog, and we have been testing and using that catalog and installing it as an app from Rancher UI. We have continued testing with nodeport enabled and disabled, clusterip works as well.
 
**Checklist**
<!-- [Place an '[X]' (no spaces) in all applicable fields. Please remove unrelated fields.] -->
- [X] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [X] Variables are documented in the README.md
- [X] Title of the PR starts with chart name (e.g. [bitnami/<name_of_the_chart>])

KeithTt commented:
Hi @Maelstrom96, how to enable nodePort for sentinel, I set `sentinel.service.nodePorts.redis:30009` and `sentinel.service.nodePorts.sentinel:30010`, but it do not work, the service type is still `clusterIP`.

Here is my service yaml:
```yaml
---
apiVersion: v1
kind: Service
metadata:
  annotations:
    meta.helm.sh/release-name: redis-sentinel
    meta.helm.sh/release-namespace: redis
  labels:
    app.kubernetes.io/component: node
    app.kubernetes.io/instance: redis-sentinel
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: redis
    helm.sh/chart: redis-15.3.2
  name: redis-sentinel
  namespace: redis
  resourceVersion: '304985783'
spec:
  clusterIP: 10.3.120.157
  ports:
    - name: tcp-redis
      port: 6379
      protocol: TCP
      targetPort: redis
    - name: tcp-sentinel
      port: 26379
      protocol: TCP
      targetPort: redis-sentinel
  selector:
    app.kubernetes.io/component: node
    app.kubernetes.io/instance: redis-sentinel
    app.kubernetes.io/name: redis
  sessionAffinity: None
  type: ClusterIP
```

alvneiayu commented:
hi @Maelstrom96 

Thanks a lot for your PR and your time. Really good work. I need some time to test it and verify the code. I will come back as soon as possible with feedback.

Thanks a lot again
Álvaro

alvneiayu commented on bitnami/redis/Chart.yaml:
```diff
@@ -25,4 +25,4 @@ name: redis
   - https://github.com/bitnami/bitnami-docker-redis
   - http://redis.io/
-version: 15.3.2
+version: 15.3.3
```
I think that this is a minor change

Maelstrom96 replied:
Fixed: https://github.com/bitnami/charts/pull/7461/commits/2835a1e15271f260ae2754a3f02e222ed4f86981

alvneiayu commented on bitnami/redis/templates/scripts-configmap.yaml:
```diff
@@ -63,9 +123,26 @@ data:
+        retry_count=1
+        while validate_quorum
+        do
+            info "sleeping, waiting for redis master to come up"
```
Please, modify redis to Redis for the logs.

Maelstrom96 replied:
Fixed: https://github.com/bitnami/charts/pull/7461/commits/11a7111c6147ed689babc3bfe4db2a922bacf91b

alvneiayu commented on bitnami/redis/templates/scripts-configmap.yaml:
```diff
@@ -63,9 +123,26 @@ data:
+            
+            ((retry_count+=1))
+        done
+        info "redis master is up now"
```
Please, modify redis to Redis for the logs.

Maelstrom96 replied:
Fixed: https://github.com/bitnami/charts/pull/7461/commits/11a7111c6147ed689babc3bfe4db2a922bacf91b

alvneiayu reviewed (changes requested):
hi @Maelstrom96

Tested the change in our internal pipeline. This change is a minor one, so please modify the minor version. I included also several comments.

Thanks a lot for your PR and for your really hard work.

Álvaro

Maelstrom96 commented on bitnami/redis/templates/scripts-configmap.yaml:
```diff
@@ -202,13 +321,14 @@ data:
 
         # Fetches current master's host and port
         REDIS_SENTINEL_INFO=($(get_sentinel_master_info))
+        info "priting REDIS_SENTINEL_INFO=(${REDIS_SENTINEL_INFO[0]},${REDIS_SENTINEL_INFO[1]})" 
```
Fixed: https://github.com/bitnami/charts/pull/7461/commits/2b6f4abfd2ea199383b9053f63f82efa27be0a5b

Maelstrom96 commented on bitnami/redis/templates/sentinel/node-services.yaml:
```diff
@@ -0,0 +1,71 @@
+{{- end }}
+
+---
+apiVersion: v1
```
Fixed: https://github.com/bitnami/charts/pull/7461/commits/ab7af83858573b1f369d63a2f6c4511b094d86ad

Maelstrom96 replied:
Alright, so when tested locally, this didn't work at all. it's complaining that `apps/v1` is not a valid version for the service either:

```
PS C:\Users\alexa\Documents\GitHub\charts\bitnami\redis> helm --kubeconfig ~\preprod.kube upgrade redis-test . --namespace redis-pn --create-namespace -f ~\redis-test.yml --debug
upgrade.go:123: [debug] preparing upgrade for redis-test
Error: UPGRADE FAILED: unable to recognize "": no matches for kind "Service" in version "apps/v1"
helm.go:81: [debug] unable to recognize "": no matches for kind "Service" in version "apps/v1"
UPGRADE FAILED
main.newUpgradeCmd.func2
        /home/circleci/helm.sh/helm/cmd/helm/upgrade.go:157
github.com/spf13/cobra.(*Command).execute
        /go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850
github.com/spf13/cobra.(*Command).ExecuteC
        /go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958
github.com/spf13/cobra.(*Command).Execute
        /go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
main.main
        /home/circleci/helm.sh/helm/cmd/helm/helm.go:80
runtime.main
        /usr/local/go/src/runtime/proc.go:204
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1374
```

Looking at the other service definitions, they are also explicitly using "v1".

Maelstrom96 replied:
Reverted https://github.com/bitnami/charts/commit/ab7af83858573b1f369d63a2f6c4511b094d86ad: https://github.com/bitnami/charts/pull/7461/commits/4260d763c281e914c82b0f740738dd75ddbdbe83

alvneiayu replied:
oh, you are right @Maelstrom96. Sorry, my bad. Everything is ok then.

Maelstrom96 commented on bitnami/redis/templates/sentinel/ports-configmap.yaml:
```diff
@@ -0,0 +1,96 @@
+#chosenports {{$chosenports}}}}
+
+---
+apiVersion: v1
```
Fixed: https://github.com/bitnami/charts/pull/7461/commits/ed154f59d04a202c4411475fcc013e4d2ff5733a

alvneiayu replied:
It is ok then, thanks

Maelstrom96 commented on bitnami/redis/templates/scripts-configmap.yaml:
```diff
@@ -63,9 +123,26 @@ data:
+        do
+            info "sleeping, waiting for Redis master to come up"
+            sleep 1s
+            if ! ((retry_count % 11)); then
```
Should this recovery feature be behind a config value that is disabled by default? This is pretty much a hack to recover broken clusters automatically, but it could, in theory, cause split brain issues if fired while there's network issues between nodes/pods. I've implemented this because we're going to use it ourselves, and I believe this should fix https://github.com/bitnami/charts/issues/7283 but in most cases, failover should be left to Sentinel's internal logic, and the cluster manually recovered by an operator that triggers the failover in the cluster using the sentinel command.

alvneiayu replied:
I think that it is a good idea and the users can enable/disable when they want.

Maelstrom96 replied:
I've added a new value to enable that feature.

Maelstrom96 commented:
Hi @alvneiayu,

Thank you for the review. I've updated everything in regard to your comments. We've just noticed a bug in port-configmap.yaml so please hold off on merging this PR until we can push the fix for it.

Also, I've added a comment that I would like you to check out regarding the automatic recovery feature that we've implemented.

Maelstrom96 commented:
Hi again, 

I believe this is ready to be merged.

alvneiayu commented:
hi @Maelstrom96 

LGTM but you are having linter errors, could you fix the problems this problem please? Moreover, could you include the new variable inside the README, please?.

Thanks a lot, you are really close.

Álvaro

Maelstrom96 commented:
Hi  @alvneiayu

Lint is fixed and I've added the new var to README.

alvneiayu reviewed (approved):
LGTM, could you fix the conflict? I don't have permission in your branch to fix it.

Thanks a lot and really hard work

Álvaro

Maelstrom96 commented:
@alvneiayu Hi,

The conflict has been fixed.

aeimer commented:
@Maelstrom96 The version update from `15.3.3` to `15.4.0`, so exactly that PR, makes our sentinel setup fail! Can you explain what is happening here and what we need to adapt?
The latest redisson versions are not working properly.
Also it only seems to be a problem with the Slaves, I don't see any problems with the redis in these logs?!

We have the helm chart deployed as master-slave with sentinel.

```
{"timestamp":"2021-12-07T13:51:37.729Z","level":"INFO","thread":"SimpleAsyncTaskExecutor-27","logger":"org.apache.catalina.core.StandardService","message":"Starting service [Tomcat[]"}
{"timestamp":"2021-12-07T13:51:37.730Z","level":"INFO","thread":"SimpleAsyncTaskExecutor-27","logger":"org.apache.catalina.core.StandardEngine","message":"Starting Servlet engine: [Apache Tomcat/9.0.55]"}
{"timestamp":"2021-12-07T13:51:37.738Z","level":"INFO","thread":"SimpleAsyncTaskExecutor-27","logger":"org.apache.catalina.core.ContainerBase.[Tomcat-1[].[localhost[].[/]","message":"Initializing Spring embedded WebApplicationContext"}
{"timestamp":"2021-12-07T13:51:37.738Z","level":"INFO","thread":"SimpleAsyncTaskExecutor-27","logger":"org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext","message":"Root WebApplicationContext: initialization completed in 1504 ms"}
{"timestamp":"2021-12-07T13:51:37.930Z","level":"INFO","thread":"SimpleAsyncTaskExecutor-27","logger":"org.springframework.boot.actuate.endpoint.web.EndpointLinksResolver","message":"Exposing 3 endpoint(s) beneath base path '/actuator'"}
{"timestamp":"2021-12-07T13:51:38.430Z","level":"INFO","thread":"SimpleAsyncTaskExecutor-27","logger":"org.apache.coyote.http11.Http11NioProtocol","message":"Starting ProtocolHandler [\"http-nio-8081\"]"}
{"timestamp":"2021-12-07T13:51:38.433Z","level":"INFO","thread":"SimpleAsyncTaskExecutor-27","logger":"org.springframework.boot.web.embedded.tomcat.TomcatWebServer","message":"Tomcat started on port(s): 8081 (http) with context path ''"}
{"timestamp":"2021-12-07T13:51:44.132Z","level":"INFO","thread":"http-nio-8081-exec-1","logger":"org.apache.catalina.core.ContainerBase.[Tomcat-1[].[localhost[].[/]","message":"Initializing Spring DispatcherServlet 'dispatcherServlet'"}
{"timestamp":"2021-12-07T13:51:44.133Z","level":"INFO","thread":"http-nio-8081-exec-1","logger":"org.springframework.web.servlet.DispatcherServlet","message":"Initializing Servlet 'dispatcherServlet'"}
{"timestamp":"2021-12-07T13:51:44.134Z","level":"INFO","thread":"http-nio-8081-exec-1","logger":"org.springframework.web.servlet.DispatcherServlet","message":"Completed initialization in 1 ms"}
{"timestamp":"2021-12-07T13:52:02.258Z","level":"WARN","thread":"redisson-netty-2-13","logger":"org.redisson.connection.SentinelConnectionManager","message":"slave: redis://10.0.16.188:6379 is down"}
{"timestamp":"2021-12-07T13:52:04.637Z","level":"INFO","thread":"redisson-netty-2-20","logger":"org.redisson.connection.SentinelConnectionManager","message":"slave: redis://10.0.16.188:6379 is up"}
{"timestamp":"2021-12-07T13:52:04.638Z","level":"INFO","thread":"redisson-netty-2-15","logger":"org.redisson.connection.pool.PubSubConnectionPool","message":"1 connections initialized for 10.0.16.188/10.0.16.188:6379"}
{"timestamp":"2021-12-07T13:52:04.741Z","level":"INFO","thread":"redisson-netty-2-28","logger":"org.redisson.connection.pool.SlaveConnectionPool","message":"24 connections initialized for 10.0.16.188/10.0.16.188:6379"}
{"timestamp":"2021-12-07T13:52:12.666Z","level":"WARN","thread":"redisson-netty-2-12","logger":"org.redisson.connection.SentinelConnectionManager","message":"slave: redis://10.0.16.188:6379 is down"}
{"timestamp":"2021-12-07T13:52:13.668Z","level":"INFO","thread":"redisson-netty-2-32","logger":"org.redisson.connection.SentinelConnectionManager","message":"slave: redis://10.0.16.188:6379 is up"}
{"timestamp":"2021-12-07T13:52:13.673Z","level":"INFO","thread":"redisson-netty-2-26","logger":"org.redisson.connection.pool.PubSubConnectionPool","message":"1 connections initialized for 10.0.16.188/10.0.16.188:6379"}
{"timestamp":"2021-12-07T13:52:13.677Z","level":"INFO","thread":"redisson-netty-2-28","logger":"org.redisson.connection.pool.SlaveConnectionPool","message":"24 connections initialized for 10.0.16.188/10.0.16.188:6379"}
{"timestamp":"2021-12-07T13:52:26.896Z","level":"WARN","thread":"redisson-netty-2-14","logger":"org.redisson.connection.SentinelConnectionManager","message":"slave: redis://10.0.16.188:6379 is down"}
{"timestamp":"2021-12-07T13:52:27.897Z","level":"INFO","thread":"redisson-netty-2-20","logger":"org.redisson.connection.SentinelConnectionManager","message":"slave: redis://10.0.16.188:6379 is up"}
{"timestamp":"2021-12-07T13:52:27.902Z","level":"INFO","thread":"redisson-netty-2-3","logger":"org.redisson.connection.pool.PubSubConnectionPool","message":"1 connections initialized for 10.0.16.188/10.0.16.188:6379"}
{"timestamp":"2021-12-07T13:52:27.903Z","level":"INFO","thread":"redisson-netty-2-13","logger":"org.redisson.connection.pool.SlaveConnectionPool","message":"24 connections initialized for 10.0.16.188/10.0.16.188:6379"}
{"timestamp":"2021-12-07T13:52:54.965Z","level":"WARN","thread":"redisson-netty-2-9","logger":"org.redisson.connection.SentinelConnectionManager","message":"slave: redis://10.0.16.188:6379 is down"}
[...]
```

juan131 commented:
@aeimer we're tracking the problems related to Redis sentinel mode in the issue below:

- https://github.com/bitnami/charts/issues/7283

Please continue the conversation in that thread.
